fineract-provider/src/main/java/org/apache/fineract/infrastructure/event/external/service/ExternalEventConfigurationValidationService.java

```diff
@@ -0,0 +1,96 @@
+import io.github.classgraph.ClassInfoList;
+import io.github.classgraph.ScanResult;
+import java.util.List;
+import java.util.stream.Collectors;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.fineract.infrastructure.core.domain.FineractPlatformTenant;
+import org.apache.fineract.infrastructure.core.service.ThreadLocalContextUtil;
+import org.apache.fineract.infrastructure.event.external.exception.ExternalEventConfigurationNotFoundException;
+import org.apache.fineract.infrastructure.event.external.repository.ExternalEventConfigurationRepository;
+import org.apache.fineract.infrastructure.security.service.TenantDetailsService;
+import org.springframework.beans.factory.InitializingBean;
+import org.springframework.stereotype.Service;
+
+@Slf4j
+@RequiredArgsConstructor
+@Service
+public class ExternalEventConfigurationValidationService implements InitializingBean {
+
+    private static final String EXTERNAL_EVENT_CLASSES_BASE_PACKAGE = "org.apache.fineract.infrastructure.event.business.domain";
+    private static final String EXTERNAL_EVENT_BUSINESS_INTERFACE = "org.apache.fineract.infrastructure.event.business.domain.BusinessEvent";
+    private static final String BULK_BUSINESS_EVENT = "org.apache.fineract.infrastructure.event.business.domain.BulkBusinessEvent";
+    private final TenantDetailsService tenantDetailsService;
+    private final ExternalEventConfigurationRepository repository;
+
+    @Override
+    public void afterPropertiesSet() throws Exception {
+        validateEventConfigurationForAllTenants();
+    }
+
+    private void validateEventConfigurationForAllTenants() throws ExternalEventConfigurationNotFoundException {
+        List<String> eventClasses = getAllEventClasses();
+        List<FineractPlatformTenant> tenants = tenantDetailsService.findAllTenants();
+
+        if (isNotEmpty(tenants)) {
+            for (FineractPlatformTenant tenant : tenants) {
+                validateEventConfigurationForIndividualTenant(tenant, eventClasses);
+            }
+        }
+    }
+
+    private void validateEventConfigurationForIndividualTenant(FineractPlatformTenant tenant, List<String> eventClasses)
+            throws ExternalEventConfigurationNotFoundException {
+        log.info("Validating external event configuration for {}", tenant.getTenantIdentifier());
+        List<String> eventConfigurations = getExternalEventConfigurationsForTenant(tenant);
+
+        if (eventClasses.size() != eventConfigurations.size()) {
+            throw new ExternalEventConfigurationNotFoundException();
+        }
+
+        for (String eventTypeClass : eventClasses) {
+            if (!eventConfigurations.contains(eventTypeClass)) {
+                throw new ExternalEventConfigurationNotFoundException(eventTypeClass);
+            }
+        }
+    }
+
+    private List<String> getExternalEventConfigurationsForTenant(FineractPlatformTenant tenant) {
+        ThreadLocalContextUtil.setTenant(tenant);
```

Comment:
I'd be careful with this since there's gonna be garbage leftover after the last iteration. Either we save the initial ThreadLocalContextUtil state and restore after we're done here or you can use the `org.apache.fineract.infrastructure.core.service.migration.TenantDataSourceFactory#create` method to get a tenant specific DataSource without manipulating the threadlocal thingy. Of course this means you get a plain DataSource from which you should construct some kind of higher level abstraction data access component (all explicitly). Thoughts?

Reply:
This is what i tried initially:
private List<String> getExternalEventConfigurationsForTenant(FineractPlatformTenant tenant) {
        DataSource tenantDataSource = tenantDataSourceFactory.create(tenant);
        final JdbcTemplate jdbcTemplate = new JdbcTemplate(tenantDataSource);
        final StringBuilder eventConfigurations = new StringBuilder();
        eventConfigurations.append("select me.type as type from m_external_event_configuration me");
        List<String> configuredEventTypes = jdbcTemplate.queryForList(eventConfigurations.toString(), String.class);
        return configuredEventTypes;
    }
Please comment .